### DISCUSSION TITLE
[v5] Async custom validation rule

### DISCUSSION THREAD

**Discussion by maxgalbu** [adonisjs/core#2912] [Help] [answered]
I'm trying to implement a "non-disponsable email" rule for adonis v5. To do this, i need to download a list of disposable domains (eg 10minutemail.com) and THEN validate emails. 

I'm following this "guide", since there's nothing in official docs: https://github.com/adonisjs/legacy.adonisjs.com/blob/develop/docs/docs/guides/validator/08-adding-custom-rules.md

```js
let intervalId = 0;
let disposableDomains = [];

validator.rule(this.ruleName, async (value, _args, {pointer, errorReporter}) => {
    await fetchDisposableDomains();

    //As soon as it's used once, start a timed retrieval of new domains.
    if (!intervalId) {
        //Refresh every hour
        intervalId = setInterval(async() => {
            await fetchDisposableDomains(); //  👈
        }, 60 * 60 * 1000);
    }

    const [_, emailDomain] = value.split("@");

    if (disposableDomains.indexOf(emailDomain) > -1) {
        errorReporter.report(pointer, this.ruleName, "You cannot use this email");
    }
});
```

Thing is, I don't think the validation.rule() function accepts an async function?

**Comment by arthurcvm** [comment, marked as answer]
Hi @maxgalbu!
Custom validators accepts async functions and this is very simple.
You can pass the `compileFn` in the rule returning the `async` option, look the example:

```ts
validator.rule('nameOfRule', async (value, _, { pointer, arrayExpressionPointer, errorReporter }) => {
  if (isNaN(value)) {
    return
  }

  const permission = await Permission.query().where('id', value).first()

  if (!permission) {
    return
  }

  if (permission.type !== 'page') {
    errorReporter.report(pointer, 'nameOfRule', 'Invalid permission type', arrayExpressionPointer)
  }
},
() => {
  return {
    async: true
  }
})
```

**Reply by maxgalbu** [reply]
oh, good!

is this documented somewhere?

**Reply by arthurcvm** [reply]
Hi @maxgalbu!
I created a [PR](https://github.com/adonisjs/legacy.adonisjs.com/pull/163) but after I created the new docs is ready.
I created another [PR](https://github.com/adonisjs/docs.adonisjs.com/pull/21), and you found, but it was not accepted.
